src/java.base/share/classes/java/io/FileInputStream.java

```diff
@@ -331,6 +331,9 @@ public int read(byte[] b, int off, int len) throws IOException {
 
     @Override
     public byte[] readAllBytes() throws IOException {
+        if (!canSeek())
```

Comment:
Since `canSeek()` is implemented as a native call whose result depends on the file descriptor of this `FileInputStream`, do you think we should reduce these native calls in this change and call `canSeek()` just once (and store that state) in the constructors of this class?
If we do that, we might have to consider whether the `FileInputStream(FileDescriptor fdObj)` will end up throwing an exception (from the native canSeek() implementation) if the `FileDescriptor` passed to the constructor is not `valid()` https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/io/FileDescriptor.html#valid()

Reply:
readAllBytes reads to EOF so there isn't any benefit to caching. I think the main thing with this PR is whether canSeek is the right thing to use.

Reply:
Hello Alan, I didn't just mean the readAllBytes. The rest of the changes in this PR calls canSeek() even in readNBytes() and skip() implementations which may not be reading till EOF and in theory could be invoked multiple times.

Reply:
> [...] do you think we should reduce these native calls in this change and call `canSeek()` just once [...]

I think that is a good idea but I would think it best to do i in a similar way to the recent change to `ChannelInputStream` that added `isOther`.

Reply:
> I think the main thing with this PR is whether canSeek is the right thing to use.

`canSeek` is really the same as `isOther` in the recent `Files.newInputStream` change, but given that on Windows the result of the function is always `true`, the naming `isOther` (which would always return `false`) did not seem apropos.

Reply:
I think it would be limit to limit is to regular files, in which case canSeek could be isRegularFile and just check S_ISREG. The Windows implementation that just assume it's a regular file until there is a better implementation. Up to you but I think it would make it much clearer. Longer time we need to re-target this to use FileChannelImpl but several issues to work through before then.

Reply:
So changed in [0bb34b4](https://github.com/openjdk/jdk/pull/21673/commits/0bb34b4d62d9ac89b9d84098a6df95b77838c724).